hedera-node/src/main/java/com/hedera/services/state/expiry/renewal/RenewalHelper.java

```diff
@@ -0,0 +1,216 @@
+	private final Supplier<FCMap<MerkleEntityId, MerkleToken>> tokens;
+	private final Supplier<FCMap<MerkleEntityId, MerkleAccount>> accounts;
+	private final Supplier<FCMap<MerkleEntityAssociation, MerkleTokenRelStatus>> tokenRels;
+
+	private MerkleAccount lastClassifiedAccount = null;
+	private MerkleEntityId lastClassifiedEntityId;
+
+	public RenewalHelper(
+			TokenStore tokenStore,
+			HederaNumbers hederaNumbers,
+			GlobalDynamicProperties dynamicProperties,
+			Supplier<FCMap<MerkleEntityId, MerkleToken>> tokens,
+			Supplier<FCMap<MerkleEntityId, MerkleAccount>> accounts,
+			Supplier<FCMap<MerkleEntityAssociation, MerkleTokenRelStatus>> tokenRels
+	) {
+		this.shard = hederaNumbers.shard();
+		this.realm = hederaNumbers.realm();
+		this.tokens = tokens;
+		this.tokenStore = tokenStore;
+		this.accounts = accounts;
+		this.tokenRels = tokenRels;
+		this.dynamicProperties = dynamicProperties;
+	}
+
+	public ExpiredEntityClassification classify(long candidateNum, long now) {
+		lastClassifiedEntityId = new MerkleEntityId(shard, realm, candidateNum);
+		var currentAccounts = accounts.get();
+
+		if (!currentAccounts.containsKey(lastClassifiedEntityId)) {
+			return OTHER;
+		} else {
+			lastClassifiedAccount = currentAccounts.get(lastClassifiedEntityId);
+
+			final long expiry = lastClassifiedAccount.getExpiry();
+			if (expiry > now) {
+				return OTHER;
+			}
+			if (lastClassifiedAccount.getBalance() > 0) {
+				return EXPIRED_ACCOUNT_READY_TO_RENEW;
+			}
+
+			final long gracePeriodEnd = expiry + dynamicProperties.autoRenewGracePeriod();
+			if (gracePeriodEnd > now) {
+				return DETACHED_ACCOUNT;
+			}
+
+			final var grpcId = lastClassifiedEntityId.toAccountId();
+			if (tokenStore.isKnownTreasury(grpcId)) {
+				return DETACHED_TREASURY_GRACE_PERIOD_OVER_BEFORE_TOKEN;
+			}
+
+			return DETACHED_ACCOUNT_GRACE_PERIOD_OVER;
+		}
+	}
+
+	public List<TokenTransferList> removeLastClassifiedAccount() {
+		assertHasLastClassifiedAccount();
+		if (lastClassifiedAccount.getBalance() > 0) {
+			throw new IllegalStateException("Cannot remove the last classified account, has non-zero balance!");
+		}
+
+		List<TokenTransferList> tokensDisplaced = Collections.emptyList();
+		final var lastClassifiedTokens = lastClassifiedAccount.tokens();
+		if (lastClassifiedTokens.numAssociations() > 0) {
+			final var grpcId = lastClassifiedEntityId.toAccountId();
+			final var currentTokens = tokens.get();
+			final List<TokenTransferList> displacements = new ArrayList<>();
+			for (var tId : lastClassifiedTokens.asIds()) {
+				doReturnToTreasury(grpcId, tId, displacements, currentTokens);
+			}
+			tokensDisplaced = displacements;
+		}
+
+		final var currentAccounts = accounts.get();
+		currentAccounts.remove(lastClassifiedEntityId);
+
+		return tokensDisplaced;
+	}
+
+	public void renewLastClassifiedWith(long fee, long renewalPeriod) {
+		assertHasLastClassifiedAccount();
+		assertLastClassifiedAccountCanAfford(fee);
```

Comment:
So, we only autoRenew if we have enough fee to renew for renewalPeriod ? I remember the design doc mentioning that we renew for what ever `time` the remaining balance can get us if its not enough for mentioned `time`